app/services/work_packages/shared/working_days.rb

```diff
@@ -0,0 +1,121 @@
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module WorkPackages
+  module Shared
+    class WorkingDays
+      # Returns number of working days between two dates, excluding weekend days
+      # and non working days.
+      def duration(start_date, due_date)
+        return no_duration unless start_date && due_date
+
+        (start_date..due_date).count { working?(_1) }
+      end
+
+      def due_date(start_date, duration)
+        return nil unless start_date && duration
+        raise ArgumentError, 'duration must be strictly positive' if duration.is_a?(Integer) && duration <= 0
+
+        due_date = start_date
+        until duration <= 1 && working?(due_date)
+          due_date += 1
+          duration -= 1 if working?(due_date)
+        end
+        due_date
+      end
+
+      def add_days(date, count)
+        while count > 0
+          date += 1
+          count -= 1 if working?(date)
+        end
+        while count < 0
+          date -= 1
+          count += 1 if working?(date)
+        end
+        date
+      end
+
+      def soonest_working_day(date)
+        return unless date
+
+        until working?(date)
+          date += 1
+        end
+        date
+      end
+
+      def delta(previous:, current:)
+        delta = 0
+        direction = previous < current ? 1 : -1
+        pos = last_pos = previous
+        while pos != current
+          pos += direction
+          if working?(last_pos) && working?(pos)
+            delta += direction
+            last_pos = pos
+          end
+        end
+        delta
+      end
+
+      def working?(date)
+        working_week_day?(date) && working_specific_date?(date)
+      end
+
+      private
+
+      def no_duration
+        OpenProject::FeatureDecisions.work_packages_duration_field_active? ? nil : 1
+      end
+
+      def working_week_day?(date)
+        working_week_days[date.wday]
+      end
+
+      def working_specific_date?(date)
+        non_working_dates.exclude?(date)
+      end
+
+      def working_week_days
+        return @working_week_days if defined?(@working_week_days)
+
+        @working_week_days = [true] * 8
+        WeekDay.pluck(:day, :working).each do |day, working|
+          @working_week_days[day] = working
+        end
+        @working_week_days[0] = @working_week_days[7] # Sunday is 7 in iso or 0 in other implementations
```

Comment:
It builds an array with information about which day is working or not. Because Sunday is sometimes on 0 index (because that's what Date#wday returns), and sometimes on 7 index (because we chose to conform to ISO), an array of size 8 is built with the information on index 0 and 7 being duplicated. So the array is [Sunday, Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday]

Because database table for week days could be empty, it builds it with `true` values everywhere first.
```
        @working_week_days = [true] * 8
```
Then picks the values from the db which will fill index 1-7
```
        WeekDay.pluck(:day, :working).each do |day, working|
          @working_week_days[day] = working
        end
```
And finally copies the value for Sunday at index 7 to the index 0
```
         @working_week_days[0] = @working_week_days[7] # Sunday is 7 in iso or 0 in other implementations
```

So with Saturday and Sunday being weekend days, the array should be 
```
[false, true, true, true, true, true, false, false]
```

Is is clearer now? I'll add some comments next to the code.

Reply:
comments added in [9568abf](https://github.com/opf/openproject/pull/10851/commits/9568abf2e279f24ae6cd8fc147f39f50e46a8e85)